pkg/azure/actuator.go

```diff
@@ -81,41 +104,480 @@ func isAzureCredentials(providerSpec *runtime.RawExtension) (bool, error) {
+		err := codec.DecodeProviderSpec(cr.Spec.ProviderSpec, &azureSpec)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding provider v1 spec: %v", err)
+		}
+		return &azureSpec, nil
+	}
+
+	return nil, fmt.Errorf("no providerSpec defined")
+}
+
+func (a *Actuator) updateProviderStatus(ctx context.Context, logger log.FieldLogger, cr *minterv1.CredentialsRequest, azureStatus *minterv1.AzureProviderStatus) error {
+	var err error
+	cr.Status.ProviderStatus, err = a.codec.EncodeProviderStatus(azureStatus)
+	if err != nil {
+		logger.WithError(err).Error("error encoding provider status")
 		return err
 	}
-	if err := a.IsValidMode(); err != nil {
+
+	if cr.Status.Conditions == nil {
+		cr.Status.Conditions = []minterv1.CredentialsRequestCondition{}
+	}
+
+	err = a.client.Status().Update(ctx, cr)
+	if err != nil {
+		logger.WithError(err).Error("error updating credentials request status")
+		return err
+	}
+	return nil
+}
+
+func (a *Actuator) getExistingClientSecret(ctx context.Context, cr *minterv1.CredentialsRequest, secretLastSyncGeneration int64) string {
+	existing := &corev1.Secret{}
+	if err := a.client.Get(ctx, client.ObjectKey{Namespace: cr.Spec.SecretRef.Namespace, Name: cr.Spec.SecretRef.Name}, existing); err == nil {
+		if existing.Generation == secretLastSyncGeneration {
+			return string(existing.Data[AzureClientSecret])
+		}
+	}
+	return ""
+}
+
+func (a *Actuator) syncMint(ctx context.Context, cr *minterv1.CredentialsRequest, cloudCredsSecret *corev1.Secret, infraName string, infraResourceGroups []string, logger log.FieldLogger) error {
+	azureStatus, err := decodeProviderStatus(a.codec, cr)
+	if err != nil {
+		return err
+	}
+
+	azureSpec, err := decodeProviderSpec(a.codec, cr)
+	if err != nil {
+		return err
+	}
+
+	targetRole := defaultTargetRole
+	if len(azureSpec.RoleBindings) > 0 {
+		targetRole = azureSpec.RoleBindings[0].Role
+	} else {
+		logger.Infof("Target role not specified, falling back to %q", defaultTargetRole)
+	}
+
+	spName := azureStatus.ServicePrincipalName
+	if spName == "" {
+		spName, err = generateServicePrincipalName(infraName, cr.Name)
+		if err != nil {
+			return err
+		}
+	}
+
+	azureCredentialsMinter, err := newAzureCredentialsMinter(
+		logger,
+		string(cloudCredsSecret.Data[AzureClientID]),
+		string(cloudCredsSecret.Data[AzureClientSecret]),
+		string(cloudCredsSecret.Data[AzureTenantID]),
+		string(cloudCredsSecret.Data[AzureSubscriptionID]),
+	)
+	if err != nil {
+		return fmt.Errorf("unable to create azure cred minter: %v", err)
+	}
+
+	// Client secret can not be retrieved from Azure, it can be only re-generated.
+	// Thus, either use already existing (if it can be found) or generate new one.
+	clientSecret := ""
+	existingSecret := &corev1.Secret{}
+	if err := a.client.Get(ctx, client.ObjectKey{Namespace: cr.Spec.SecretRef.Namespace, Name: cr.Spec.SecretRef.Name}, existingSecret); err == nil {
+		if existingSecret.ResourceVersion == azureStatus.SecretLastResourceVersion {
```

Comment:
What scenario is this meant to catch? Someone manually editing the target credentials secret? 

For AWS we do a dynamic check to ensure the credentials are still functional, I'm tempted to say we should remove this SecretLastResourceVersion field from the API and instead validate that (a) the creds are working by hitting the API with them, and (b) the appropriate roles are on the creds.

Is that feasible for this PR or something reasonable for future? The fact that it's an API change makes me think we should perhaps avoid shipping this field out if we don't want to keep it long term.